# Getting started with Metrica Sports tracking data

In this notebook, we'll use event and tracking data provided by Metrica Sports to achieve the following targets:

* Plot one of the goals scored.
* Plot the passes leading up to that goal.
* Plot the positions of all the players (from both the teams) when this goal was scored.
* Plot the movements of players involved in this goal.

### Import libraries and modules

In [1]:
import modules.Metrica_IO as mio # module for input output operations involving Metrica Sports data
import modules.Metrica_Viz as mviz # module for visualising Metrica Sports data

### Initial setup

In [2]:
DATADIR = 'LaurieOnTracking-master/sample-data-master/data/' # setting up the path to data
game_id = 1 # initialising the game id (we will be analysing the game 1's data; Metrica Sports has provided data for 2 games)

### Read in the event and the tracking data

In [3]:
events = mio.read_event_data(DATADIR,game_id) # reading in the event data
tracking_home = mio.tracking_data(DATADIR,game_id,'Home') # reading in the tracking data for the home team
tracking_away = mio.tracking_data(DATADIR,game_id,'Away') # reading in the tracking data for the away team

Reading team: home
Reading team: away


In [7]:
events.head(3)

Team       Type   Subtype  Period  Start Frame  Start Time [s]  End Frame  \
0  Away  SET PIECE  KICK OFF       1            1            0.04          0   
1  Away       PASS       NaN       1            1            0.04          3   
2  Away       PASS       NaN       1            3            0.12         17   

   End Time [s]      From        To  Start X  Start Y  End X  End Y  
0          0.00  Player19       NaN      NaN      NaN    NaN    NaN  
1          0.12  Player19  Player21     0.45     0.39   0.55   0.43  
2          0.68  Player21  Player15     0.55     0.43   0.58   0.21

In [8]:
tracking_home.head(3)

Period  Time [s]  Home_11_x  Home_11_y  Home_1_x  Home_1_y  Home_2_x  \
Frame                                                                         
1           1      0.04    0.00082    0.48238   0.32648   0.65322   0.33701   
2           1      0.08    0.00096    0.48238   0.32648   0.65322   0.33701   
3           1      0.12    0.00114    0.48238   0.32648   0.65322   0.33701   

       Home_2_y  Home_3_x  Home_3_y  ...  Home_10_x  Home_10_y  Home_12_x  \
Frame                                ...                                    
1       0.48863   0.30927   0.35529  ...    0.55243    0.43269        NaN   
2       0.48863   0.30927   0.35529  ...    0.55243    0.43269        NaN   
3       0.48863   0.30927   0.35529  ...    0.55243    0.43269        NaN   

       Home_12_y  Home_13_x  Home_13_y  Home_14_x  Home_14_y   ball_x   ball_y  
Frame                                                                           
1            NaN        NaN        NaN        NaN        NaN  0.45472  0.38709  
2            NaN        NaN        NaN        NaN        NaN  0.49645  0.40656  
3            NaN        NaN        NaN        NaN        NaN  0.53716  0.42556  

[3 rows x 32 columns]

In [9]:
tracking_away.head(3)

Period  Time [s]  Away_25_x  Away_25_y  Away_15_x  Away_15_y  \
Frame                                                                 
1           1      0.04    0.90509    0.47462    0.58393    0.20794   
2           1      0.08    0.90494    0.47462    0.58393    0.20794   
3           1      0.12    0.90434    0.47463    0.58393    0.20794   

       Away_16_x  Away_16_y  Away_17_x  Away_17_y  ...  Away_24_x  Away_24_y  \
Frame                                              ...                         
1        0.67658     0.4671     0.6731    0.76476  ...    0.37833    0.27383   
2        0.67658     0.4671     0.6731    0.76476  ...    0.37833    0.27383   
3        0.67658     0.4671     0.6731    0.76476  ...    0.37833    0.27383   

       Away_26_x  Away_26_y  Away_27_x  Away_27_y  Away_28_x  Away_28_y  \
Frame                                                                     
1            NaN        NaN        NaN        NaN        NaN        NaN   
2            NaN        NaN        NaN        NaN        NaN        NaN   
3            NaN        NaN        NaN        NaN        NaN        NaN   

        ball_x   ball_y  
Frame                    
1      0.45472  0.38709  
2      0.49645  0.40656  
3      0.53716  0.42556  

[3 rows x 32 columns]

Before any kind of analysis and visualisation, we'll convert the units of all the locations in the 3 dataframes from Metrica Sports units to metres:

In [10]:
events = mio.to_metric_coordinates(events)
tracking_home = mio.to_metric_coordinates(tracking_home)
tracking_away = mio.to_metric_coordinates(tracking_away)

### Data analysis and visualisation

First of all, let's create a dataframe for all the shots taken in the match.

In [12]:
shots = events[events['Type'] == 'SHOT']

Now, let's have a look at the data of the shots which ended up as goals.

In [13]:
shots[shots['Subtype'].str.contains('-GOAL')]

Team  Type              Subtype  Period  Start Frame  Start Time [s]  \
34    Home  SHOT  HEAD-ON TARGET-GOAL       1         2289           91.56   
1114  Home  SHOT       ON TARGET-GOAL       2        90005         3600.20   
1213  Home  SHOT       ON TARGET-GOAL       2        99032         3961.28   

      End Frame  End Time [s]      From   To  Start X  Start Y  End X  End Y  
34         2309         92.36   Player9  NaN    44.52     2.04  54.06  -3.40  
1114      90026       3601.04  Player10  NaN   -44.52    12.92 -53.00  -2.72  
1213      99046       3961.84   Player9  NaN   -50.88    -2.72 -54.06  -0.68

For further analysis and visualisation, we'll select the second goal scored in the match (with index number = 1114). Let's save that goal's index number for future use.

In [14]:
goal_index = 1114

Let's have a look at the last 10 events leading to this goal.

In [15]:
events.loc[goal_index-10:goal_index]

Team       Type             Subtype  Period  Start Frame  \
1104  Home       PASS                 NaN       2        89622   
1105  Home       PASS                 NaN       2        89691   
1106  Home       PASS                 NaN       2        89726   
1107  Home  BALL LOST  CROSS-INTERCEPTION       2        89802   
1108  Away   RECOVERY        INTERCEPTION       2        89832   
1109  Away  BALL LOST      HEAD-CLEARANCE       2        89832   
1110  Home   RECOVERY                 NaN       2        89875   
1111  Home       PASS                HEAD       2        89875   
1112  Home       PASS                 NaN       2        89955   
1113  Home       PASS                 NaN       2        89978   
1114  Home       SHOT      ON TARGET-GOAL       2        90005   

      Start Time [s]  End Frame  End Time [s]      From        To  Start X  \
1104         3584.88      89652       3586.08   Player7   Player5   -40.28   
1105         3587.64      89726       3589.04   Player5   Player7   -44.52   
1106         3589.04      89758       3590.32   Player7   Player2   -34.98   
1107         3592.08      89835       3593.40   Player2       NaN   -26.50   
1108         3593.28      89832       3593.28  Player16       NaN   -44.52   
1109         3593.28      89871       3594.84  Player16       NaN   -44.52   
1110         3595.00      89875       3595.00   Player5       NaN   -36.04   
1111         3595.00      89900       3596.00   Player5  Player10   -36.04   
1112         3598.20      89967       3598.68  Player10   Player5   -34.98   
1113         3599.12      90005       3600.20   Player5  Player10   -37.10   
1114         3600.20      90026       3601.04  Player10       NaN   -44.52   

      Start Y  End X  End Y  
1104    31.28 -42.40  17.68  
1105    21.76 -34.98  28.56  
1106    28.56 -20.14  22.44  
1107    23.12 -45.58   6.12  
1108     3.40    NaN    NaN  
1109     3.40 -36.04  21.76  
1110    21.76    NaN    NaN  
1111    21.76 -33.92  10.88  
1112    12.24 -37.10  16.32  
1113    16.32 -44.52  12.92  
1114    12.92 -53.00  -2.72

Here, we can see that the ball was recovered in the event with index number = 1110, and the passing movement to our selected goal began in the event with index number = 1111. Let's save the index number of the event in which this passing movement began for future use.

In [16]:
pass_move_start_index = 1111